### PR TITLE
Create the Kraken

### PR DESCRIPTION
**Pull request by Williangalvani**
This is my proposal of an extensions manager. It is written to work together with https://github.com/Williangalvani/BlueOS-extensions-repository

There are some unsolved problems:
 - [ ] storing logos offline

And some features that will come later
 - [ ] better handling of killing/starting extensions
 - [ ] progress bar/text for docker pull
 - [ ] automatically pick latest version
 - [ ] checking requirements
 - [ ] updating extensions
 - [ ] filtering tags for semver only?
 - [ ] better formatting for permissions/settings
 - [ ] management of "un-managed" dockers (yes, I read that)
 - [ ] cache manifest and images

The goal is to make these as incremental changes in subsequent PRs

### CODE REVIEW

File: core/frontend/src/types/kraken.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    tag: string
+}
+
+export interface ExtensionData {
+    name: string
+    description: string
+    docker: string
+    website: string
+    versions: Dictionary<Version>
+    authors?: Author[]
```
Comment (ES-Alexander) [inline core/frontend/src/types/kraken.ts:21]: It seems quite redundant (and hard to maintain consistently) for all the author data to be provided in each extension's metadata. Perhaps there's some way the author(s) can be specified at a higher level and then referred to from the extension data, which is then retrieved when creating/refreshing the manifest?

That may need to be specified in an independent global folder or database or something (separate to the extensions), with a file/entry per author and the filename/index is the reference key (maybe `organisation-author-name` like `br-willian` for easy grouping).

Alternatively each extensions repo could have a single file of authors that each of its extensions can optionally include, but that still has some redundancies when authors of independent extensions work together on an extension.

I suppose there would then need to be a `development` author or something shown by default for unmanaged extensions.

**Review by ES-Alexander — APPROVED** [review]
Nice work! :-)
I'm ok with this being merged as is, but some of the comments may be harder/less likely to change later on, so might be worth a bit of extra thought/effort now.

File: core/frontend/src/types/kraken.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+export interface Version {
+    permissions: any
+    requirements: string | null
+    tag: string
+}
+
+export interface ExtensionData {
+    name: string
```
Comment (patrickelectric) [inline core/frontend/src/types/kraken.ts:16]: Besides name we should have a string or a tag the could be used to track the extension itself. 
The reason behind that is that the extension could change names between versions, or information as website, authors and docker may change for each version.
Maybe something like:
```typescript
export interface Extension {
    identifier: string // format: extension_owner_name.owner_extension_name, E.g: patrickelectric.tainhacpp
    data: ExtensionData[]
}

export interface ExtensionData {
    name: string
    description: string
    docker: string
    website: string
    version: Version
    authors?: Author[]
    readme?: string
    support: string
    docs?: string
    icon?: string
}
```
We could also break this in a list of Extension per version, but that would make things ugly (Like Arduino does), and do the proper filter and menu creation based on the json processing.

File: core/frontend/src/views/ExtensionManagerView.vue [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  methods: {
+    async fetchManifest(): Promise<void> {
+      await back_axios({
+        method: 'get',
+        url: `${API_URL}/extensions_manifest`,
+        timeout: 3000,
+      })
+        .then((response) => {
+          const manifest = response.data
+          this.manifest = manifest
```
Comment (patrickelectric) [inline core/frontend/src/views/ExtensionManagerView.vue:141]: Is there a need for the intermediary variable ?

File: core/frontend/src/views/ExtensionManagerView.vue [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        })
+    },
+    installedVersion(): string | undefined {
+      const extension_name = this.selected_extension?.docker
+      if (!extension_name) {
+        return undefined
+      }
+      const versions = this.installed_extensions
+        .find((extension) => extension.name === extension_name)?.versions
+      return versions?.length ? versions[0].tag : undefined
```
Comment (patrickelectric) [inline core/frontend/src/views/ExtensionManagerView.vue:202]: you can use cosmos `first` function
```suggestion
      return versions?.first()?.tag
```

File: core/services/kraken/kraken.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        await self.client.images.pull(extension.fullname())
+        container = await self.client.containers.create_or_replace(name=extension.container_name(), config=config)  # type: ignore
+        await container.start()
+
+    async def check(self, extension: Extension) -> None:
+        if not any(container["Names"][0][1:] == extension.container_name() for container in self.running_containers):
+            await self.start_extension(extension)
+
+    def load_settings(self) -> None:
+        self.manager = Manager("Kraken", SettingsV1)
```
Comment (patrickelectric) [inline core/services/kraken/kraken.py:47]: It would be nice if we could use `SERVICE_NAME`

File: core/services/kraken/main.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+app = FastAPI(
+    title="Kraken API",
+    description="Kraken is the BlueOS service responsible for installing and managing thirdy-party extensions.",
+)
+app.router.route_class = GenericErrorHandlingRoute
+logger.info("Releasing the Kraken!")
+
+
+@app.get("/extensions_manifest", status_code=status.HTTP_200_OK)
+# @temporary_cache(timeout_seconds=300)
```
Comment (patrickelectric) [inline core/services/kraken/main.py:47]: Lost code

**Review by patrickelectric — CHANGES_REQUESTED** [review]

**Comment by Williangalvani** [comment]
inheritance in manifest file
use docker tags for versions
version a lot of stuff.

File: core/services/kraken/main.py (hunk 6 above)
Comment (patrickelectric) [reply core/services/kraken/main.py:47]: You can try: https://github.com/long2ice/fastapi-cache

File: core/services/kraken/kraken.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+import aiodocker
+import aiohttp
+from aiodocker.docker import DockerContainer
+from commonwealth.settings.manager import Manager
+from loguru import logger
+
+from settings import Extension, SettingsV1
+
+REPO_URL = "https://raw.githubusercontent.com/Williangalvani/BlueOS-Extensions-Repository/gh-pages/manifest.json"
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/kraken/kraken.py:13]: yes, I'll move that over once this is accepted, before merging

File: core/frontend/src/types/kraken.ts (hunk 1 above)
Comment (Williangalvani) [reply core/frontend/src/types/kraken.ts:21]: I think we'll get there eventually, I'm trying to keep things simple. but for the final store with bazaar, we'll have a proper database.

File: core/frontend/src/components/kraken/ExtensionModal.vue [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        return 'No readme available'
+      }
+      return marked(this.extension.readme, { sanitize: true })
+    },
+    available_tags(): string[] {
+      if (this.extension && this.extension.versions) {
+        const keys = Object.keys(this.extension.versions)
+        return keys ?? []
+      }
+      return []
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/frontend/src/components/kraken/ExtensionModal.vue:158]: lol, nice.

File: core/frontend/src/views/ExtensionManagerView.vue (hunk 3 above)
Comment (Williangalvani) [reply core/frontend/src/views/ExtensionManagerView.vue:141]: nope, consider it gone.

File: core/frontend/src/views/ExtensionManagerView.vue (hunk 4 above)
Comment (Williangalvani) [reply core/frontend/src/views/ExtensionManagerView.vue:202]: nice

File: core/frontend/src/types/kraken.ts (hunk 2 above)
Comment (Williangalvani) [reply core/frontend/src/types/kraken.ts:28]: I did it as we discussed in the call, and sneaked the identifier in there, too.

File: core/services/kraken/main.py (hunk 6 above)
Comment (Williangalvani) [reply core/services/kraken/main.py]: it turned out there is a lot of overhead to this approach. I'll get a cache in there later, it is not that important

File: core/services/kraken/kraken.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    async def start_extension(self, extension: Extension) -> None:
+        config = extension.settings()
+        config["Image"] = extension.fullname()
+        logger.info(f"Starting extension '{extension.fullname()}'")
+        await self.client.images.pull(extension.fullname())
+        container = await self.client.containers.create_or_replace(name=extension.container_name(), config=config)  # type: ignore
+        await container.start()
+
+    async def check(self, extension: Extension) -> None:
+        if not any(container["Names"][0][1:] == extension.container_name() for container in self.running_containers):
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/kraken/kraken.py]: added a comment explaining it.

File: core/services/kraken/kraken.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        logger.info(f"uninstalling: {extension}")
+        if extension:
+            await self.kill(extension[0].container_name())
+        self.settings.extensions = [
+            extension for extension in self.settings.extensions if extension.name != extension_name
+        ]
+        self.manager.save()
+
+    async def list_containers(self) -> List[DockerContainer]:
+        self.client = aiodocker.Docker()
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/kraken/kraken.py]: fixed, this was a hack because it couldnt be in the init for async reasons.

**Review by ES-Alexander — APPROVED** [review]
Looks like a fantastic start! :D
Super keen to have this available so people can start making and using a variety of extensions. The structure so far seems sound and well thought out, which should make it decent to build out as we add extra features and functionalities :-)

File: core/frontend/src/components/kraken/ExtensionModal.vue [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+      if (versions && this.selected_version in versions) {
+        return versions[this.selected_version].permissions
+      }
+      return 'No permissions required'
+    },
+  },
+  watch: {
+    extension() {
+      const [first] = Object.keys(this.extension.versions) ?? ''
+      this.selected_version = first
```
Comment (patrickelectric) [inline core/frontend/src/components/kraken/ExtensionModal.vue:192]: You can unpack directly on the variable 
```suggestion
[this.selected_version] = Object.keys(this.extension.versions) ?? ''
```

File: core/frontend/src/views/ExtensionManagerView.vue [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        })
+    },
+    async fetchInstalledExtensions(): Promise<void> {
+      await back_axios({
+        method: 'get',
+        url: `${API_URL}/installed_extensions`,
+        timeout: 30000,
+      })
+        .then((response) => {
+          const extensions = response.data
```
Comment (patrickelectric) [inline core/frontend/src/views/ExtensionManagerView.vue:153]: is the intermediary variable necessary ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Just some minor corrections that should not hold the PR.
But, having it available without pirate mode is a bit dangerous, since we have not moved everything to the Blue Robotics organization, it'll be available on the next beta and also for our users without enabling pirate mode.
I'm ok-ish to have on master under pirate mode, but maybe having a branch for development would be better.

@bluerobotics/software-team What you think ?

File: core/frontend/src/components/kraken/ExtensionModal.vue (hunk 11 above)
Comment (Williangalvani) [reply core/frontend/src/components/kraken/ExtensionModal.vue:192]: this doesnt work for some reason...

```
ERROR in [eslint] 
/home/will/sambashare/companion-docker/core/frontend/src/components/kraken/ExtensionModal.vue
  191:13  error  Parsing error: Array element destructuring pattern expected


```

File: core/frontend/src/views/ExtensionManagerView.vue (hunk 12 above)
Comment (Williangalvani) [reply core/frontend/src/views/ExtensionManagerView.vue]: gone

**Review by patrickelectric — APPROVED** [review]

**Comment by patrickelectric** [comment]
LGTM
image

**Pull request merged**